---
title: "Identify Tag SNPs for Missing SNPs"
author: "Gen Roberts"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    df_print: paged
  html_notebook:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include=FALSE}
library(rstudioapi)
current_file <- rstudioapi::getActiveDocumentContext()$path
here = dirname(current_file)
library(tidyverse, lib.loc = "/share/pkg/conda/r/4.2.2/lib/R/library")
library(data.table)

#install.packages('rsnps', repos = c('https://ropensci.r-universe.dev', 'https://cloud.r-project.org'))
library(rsnps)
library(LDlinkR)

#to write the PLINK command at the end
library(glue)
library(fs)
```

## Read in Initial Weights File & PLINK .bim File

```{r}
weights_path <- str_c(here, "/Vitiligo_Risk_Score_Weights_Roberts2019.csv")
weights <- fread(weights_path)
snps <- weights$SNP

bim_path <- "/pi/manuel.garber-umw/human/VIGOR/tj/ibd/all_samples_qc.bim"
bim <- fread(bim_path)
```

## Attempt to identify synonyms for SNPs missing from the .bim file
This uses `rsnps::missing_snp_query` to see if we can pull a synonym rsID from
the same position as any missing SNP in the .bim file

```{r}
update_weights_with_rescued_snps <- function(weights, bim) {
  
  # Step 1: Identify SNPs in the weights table that are missing in the .bim table
  snps <- weights$SNP
  bim_missing_snps <- setdiff(snps, bim$V2)
  
  cat("Identified", length(bim_missing_snps), "weight SNPs not present in the .bim file. Attempting to identify synonyms...\n")
  
  # Step 2: Query missing SNPs to find chromosome_position and see if there is a synonym rsID at that position in the bim file
  missing_snp_query <- ncbi_snp_query(bim_missing_snps) %>%
    mutate(chr_pos = str_c(chromosome, ":", bp))

  cat("SNP query completed. Retrieved information for", nrow(missing_snp_query), "SNPs.\n")

  # Step 3: Attempt to rescue SNPs by joining with the .bim file
  bim <- bim %>% mutate(chr_pos=str_c(V1, ":", V4))
  
  rescue_snps <- left_join(missing_snp_query, bim, by = "chr_pos") %>%
    filter(!is.na(V1)) %>%
    transmute(
      SNP = query,
      rescue_snp = V2
    )

  cat("Successfully identified", nrow(rescue_snps), "SNP synonyms, which are now updated in the weights file.\n")
  
  # Step 4: Merge rescue SNPs back with weights and update columns
  updated_weights <- weights %>%
    left_join(rescue_snps, by = "SNP") %>%
    mutate(
      synonym_SNP = if_else(!is.na(rescue_snp), str_c(synonym_SNP, SNP, sep=","), synonym_SNP),
      SNP = if_else(!is.na(rescue_snp), rescue_snp, SNP)
    ) %>%
    select(-rescue_snp)

  return(updated_weights)
}

updated_weights <- update_weights_with_rescued_snps(weights, bim)
updated_weights
```


## Get LD Proxy SNPs for Missing SNPs
This step uses `LDlinkR` to identify LD tags for SNPs that are missing from the
PLINK .bim file for the VIGOR participants

```{r}

bim_missing_snps <- setdiff(updated_weights$SNP, bim$V2)

get_best_proxy_snps <- function(query_snps, token = "342bb55e0297", bim) {
  # Ensure query_snp is a list if it's not
  if (length(query_snps) == 1) {
    query_snps <- list(query_snps)
  }
  
  # Create an empty list to store the best proxy results for each query SNP
  best_proxies <- list()
  
  # Loop through each query SNP
  for (query_snp in query_snps) {
    # Get proxy SNPs using LDproxy function
    proxy_snps <- LDproxy(query_snp, token = token)
    
    # Find the best proxy SNP
    best_proxy <- proxy_snps %>%
      mutate(SNP = query_snp) %>%
      mutate(in_bim = RS_Number %in% bim$V2) %>%
      filter(in_bim == TRUE) %>%
      arrange(desc(R2), desc(Dprime)) %>%
      slice(1) %>%
      select(SNP, proxy_SNP = RS_Number, Alleles, Correlated_Alleles, R2, Dprime, MAF) %>%
      tibble()
    
    # Add the best proxy to the list
    best_proxies[[query_snp]] <- best_proxy
  }
  
  # Combine all results into a single dataframe
  best_proxies_df <- bind_rows(best_proxies)
  
  return(best_proxies_df)
}

best_proxies <- get_best_proxy_snps(query_snps=bim_missing_snps,
                                   token ="342bb55e0297",
                                   bim=bim)
best_proxies
```

This part ensures that, for each proxy SNP, we identify the proxy SNP allele correlated with
the original SNP's risk allele.

```{r}
process_proxy_data <- function(weights, best_proxies, effect_allele_col) {
  
  updated_table <- weights %>%
    left_join(best_proxies, by = c("SNP")) %>%
    mutate(
      effect_allele = !!sym(effect_allele_col),
      tag_snp_effect_allele = case_when(
        effect_allele == "A" ~ str_extract(Correlated_Alleles, "(?<=A=)[A-Za-z]+"),
        effect_allele == "T" ~ str_extract(Correlated_Alleles, "(?<=T=)[A-Za-z]+"),
        effect_allele == "G" ~ str_extract(Correlated_Alleles, "(?<=G=)[A-Za-z]+"),
        effect_allele == "C" ~ str_extract(Correlated_Alleles, "(?<=C=)[A-Za-z]+"),
        effect_allele == "D" ~ str_extract(Correlated_Alleles, "(?<=-=)[A-Za-z]+"),
        effect_allele == "I" ~ str_extract(Correlated_Alleles, "(?<=([A-Za-z])=)[A-Za-z]+"),
        TRUE ~ NA_character_
      ),
      other_proxy_allele = case_when(
        tag_snp_effect_allele %in% c("I", "D") ~ ifelse(
          tag_snp_effect_allele == "I", "D", "I"
        ),
        !is.na(effect_allele) & !is.na(Alleles) ~ 
          Alleles %>%
          str_remove_all("\\(|\\)|/") %>% # Remove parentheses and slash
          str_remove(tag_snp_effect_allele), # Remove the effect allele
        TRUE ~ NA_character_
      ),
      # Create proxy_snp summary column with annotations
      proxy_snp_summary = if_else(!is.na(proxy_SNP), paste(
        "Index_SNP:", SNP,
        "Proxy_SNP:", proxy_SNP,
        "R2:", R2,
        "Dprime:", Dprime,
        "MAF:", MAF,
        "Correlated_Alleles:", Correlated_Alleles
      ), NA_character_),
    ) %>%
    # Conditionally update SNP only if proxy_SNP is not NA
    mutate(SNP = if_else(!is.na(proxy_SNP), proxy_SNP, SNP)) %>%
    mutate(RA_OA = if_else(!is.na(proxy_SNP), str_c(tag_snp_effect_allele, "/", other_proxy_allele), RA_OA)) %>%
    mutate(!!effect_allele_col := if_else(!is.na(proxy_SNP), tag_snp_effect_allele, !!sym(effect_allele_col))) %>%
    # Select desired columns and drop redundant ones
    select(-proxy_SNP, -R2, -Dprime, -MAF,
           -Correlated_Alleles,
           -tag_snp_effect_allele, -other_proxy_allele, -Alleles)

  return(updated_table)
}

updated_table <- process_proxy_data(
  weights, 
  best_proxies, 
  "EA_RiskAlleles"
)
updated_table
```

This writes out the updated file that we can use to extract the SNPs that we
need from PLINK.
```{r}
#write out the new table
write.table(updated_table,
          str_c(here, "/Vitiligo_Risk_Score_weights_Roberts2019_TagSNPs_Added.tsv"),
          sep="\t",
          row.names = FALSE,
          quote = FALSE)

write.table(select(updated_table, "SNP"),
          str_c(here, "/PLINK_extract_SNPs.tsv"),
          sep="\t",
          row.names = FALSE,
          quote=FALSE)
```

## Write PLINK Command to Extract SNPs
```{r}

bfile_name = str_remove(bim_path, ".bim")
extract_file = str_c(here, "/PLINK_extract_SNPs.tsv")
output_prefix = "Roberts19_Risk_Score_SNPs_with_Tags"
output_bash_path = str_c(here, "/Extract_and_Recode_PRS_SNPs_with_Plink.sh")

bash_script <- c(
  "#!/bin/bash",
  "",
  "# Run PLINK command",
  "plink \\",
  paste0("  --bfile \"", bfile_name, "\" \\"),
  paste0("  --extract \"", extract_file, "\" \\"),
  "  --recode A \\",
  paste0("  --out \"", output_prefix, "\"")
)

# Ensure the output directory exists
writeLines(bash_script, con = output_bash_path)

# Write the Bash script using writeLines()
writeLines(bash_script, output_bash_path)

# Make the Bash script executable
Sys.chmod(output_bash_path, mode = "755")
```

