---
title: "Validate Roberts 2019 Vitiligo PRS"
author: "Gen Roberts"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    df_print: paged
  html_notebook:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(rstudioapi)
current_file <- rstudioapi::getActiveDocumentContext()$path
here = dirname(current_file)
repo_dir <- normalizePath(file.path(dirname(current_file), ".."))
library(tidyverse, lib.loc = "/share/pkg/conda/r/4.2.2/lib/R/library")
library(data.table)

#install.packages('rsnps', repos = c('https://ropensci.r-universe.dev', 'https://cloud.r-project.org'))
library(rsnps)

#to write the PLINK command at the end
library(glue)
library(fs)
```

#get the MAFs for all the SNPs

```{r}

#read in weights
weights_path <- str_c(repo_dir,
                      "/0_Extract_SNPs/Vitiligo_Risk_Score_weights_Roberts2019_TagSNPs_Added.tsv")
weights <- fread(weights_path)
snps <- weights %>%
  filter(str_starts(SNP, "rs")) %>%
  pull(SNP)
```

```{r}
#get target snp MAFs from NCBI
snp_ncbi_query <- ncbi_snp_query(snps) %>%
  transmute(SNP = query,
            chr = chromosome,
            bp,
            ancestral_allele, 
            maf_population) %>%
  unnest(cols = maf_population, names_repair = "unique") %>%
  rename(major_allele = ref_seq,
         minor_allele = Minor, 
         maf = MAF) %>%
  group_by(SNP) %>%
  arrange(SNP, 
          desc(study %in% c("TOPMED", "ExAC", "HapMap", "1000Genomes")), 
          desc(maf), 
          .by_group = TRUE) %>%
  slice_head(n = 1) %>%
  mutate(
    temp_allele = major_allele,
    major_allele = if_else(maf > 0.5, minor_allele, major_allele),
    minor_allele = if_else(maf > 0.5, temp_allele, minor_allele),
    maf=if_else(maf > 0.5, 1-maf, maf)
  ) %>%
  select(-temp_allele) %>%
  ungroup() 

snp_ncbi_query

```

```{r}

geno_dat_path <- str_c(repo_dir, "/data/Roberts19_Risk_Score_SNPs_with_Tags.raw")
geno_dat <- fread(geno_dat_path)

plink_allele_frequencies <- geno_dat %>%
  mutate(across(everything(), as.numeric)) %>%
  select(starts_with("rs")) %>%
  pivot_longer(cols = everything(), names_to = "SNPtmp", values_to = "genotype") %>%
  group_by(SNPtmp) %>%
  summarize(sum=sum(genotype), len=length(genotype)*2) %>%
  ungroup() %>%
  mutate(
    SNP = str_split_fixed(SNPtmp, "_", 2)[, 1],  # Extract the SNP ID
    minor_allele = str_split_fixed(SNPtmp, "_", 2)[, 2],  # Extract the minor allele info
    maf = sum / len
  ) %>%
  select(SNP, minor_allele, maf)

plink_allele_frequencies

```


```{r}
final_join = snp_ncbi_query %>%
  left_join(plink_allele_frequencies, by = c("SNP"), suffix = c(".ncbi", ".plink")) %>%
  left_join(weights, by = c("SNP"), suffix = c("", ".weights")) %>%
  mutate(
    maf_diff = abs(maf.ncbi - maf.plink),
    allele_mismatch = minor_allele.ncbi != minor_allele.plink
  ) %>%
  arrange(desc(maf_diff)) %>%
  select(
    SNP,
    ancestral_allele,
    major_allele.ncbi = major_allele,
    minor_allele.ncbi,
    #minor_allele.weights = minor_allele,
    minor_allele.plink,
    maf.ncbi,
    maf.plink,
    maf_diff,
    allele_mismatch,
    proxy_snp_summary
  )

final_join 

```

```{r}
ggplot(final_join, aes(x = maf.ncbi, y = maf.plink, color = allele_mismatch)) +
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "gray50", size = 1) +
  geom_point(alpha = 0.7, size = 3) +
  labs(
    title = "Scatterplot of MAFs from NCBI and PLINK",
    x = "Minor Allele Frequency (NCBI)",
    y = "Minor Allele Frequency (PLINK)",
    color = "Allele Mismatch"
  ) +
  theme_bw() +
  theme(
    text = element_text(size = 14),
    legend.position = "top"
  )

```

This plot tells us that there is reasonable-ish alignment between the MAFs estimated
from our data and the MAFs from NCBI (mostly EUR-dominated cohorts, like TOPMED).
We're mainly only seeing potential allele mismatches at MAF around 0.5, which
probably just reflects a little bit of variability in the MAF.

This is great, but **doesn't** confirm that we're scoring the correct allele since the 
risk allele is the weights allele and not necessarily the minor allele (why did we do that??).

In any case, we could simulate data for an average-looking PRS based on these
population minor allele frequencies from NCBI and make sure that the distribution
looks like the distribution of the PRS in controls. Although, I feel like there has
to be a better way to verify

